controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -274,6 +291,151 @@ func (w *Workload) UpdateKubernetesVersionInKubeadmConfigMap(ctx context.Context
+		// do nothing if it is not CoreDNS
+		// TODO: (wfernandes) Do we need this check here? If so, log something.
+		return nil
+	}
+	configMapKey := ctrlclient.ObjectKey{Name: "kubeadm-config", Namespace: metav1.NamespaceSystem}
+	kubeadmConfigMap, err := w.getConfigMap(ctx, configMapKey)
+	if err != nil {
+		return err
+	}
+	config := &kubeadmConfig{ConfigMap: kubeadmConfigMap}
+	if err := config.UpdateCoreDNSImageInfo(dns.ImageRepository, dns.ImageTag); err != nil {
+		return err
+	}
+	if err := w.Client.Update(ctx, config.ConfigMap); err != nil {
+		return errors.Wrap(err, "error updating kubeadm ConfigMap")
+	}
+	return nil
+}
+
+// TODO: (wfernandes) Add some logs
+// UpdateCoreDNSCorefile migrates the coredns corefile if there is an increase
+// in version number. It also creates a corefile backup and patches the
+// deployment to point to the backup corefile before migrating.
+func (w *Workload) UpdateCoreDNSCorefile(ctx context.Context, dns *kubeadmv1.DNS) error {
+	if dns.Type != kubeadmv1.CoreDNS {
+		// do nothing if it is not CoreDNS
+		return nil
+	}
+
+	currentCoreDNSVersion, currentCoreDNSCorefile, err := w.GetCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	toCoreDNSVersion := dns.ImageTag
+
+	if err := w.Client.Update(ctx, &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			corefileKey:       currentCoreDNSCorefile,
+			corefileBackupKey: currentCoreDNSCorefile,
+		},
+	}); err != nil {
+		return errors.Wrap(err, "unable to update CoreDNS config map with backup Corefile")
+	}
+
+	patchJSON := fmt.Sprintf(`{"spec":{"template":{"spec":{"volumes":[{"name": "config-volume", "configMap":{"name": "coredns", "items":[{"key": "%[1]s", "path": "%[1]s"}]}}]}}}}`, corefileBackupKey)
+	if err := w.patchCoreDNSDeployment(ctx, patchJSON); err != nil {
+		return errors.Wrapf(err, "unable to patch coredns deployment with corefile: %s", corefileBackupKey)
+	}
+
+	updatedCorefile, err := w.Migrator.Migrate(currentCoreDNSVersion, toCoreDNSVersion, currentCoreDNSCorefile, false)
+	if err != nil {
+		return errors.Wrap(err, "unable to migrate CoreDNS corefile")
+	}
+	if err := w.Client.Update(ctx, &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			"Corefile":        updatedCorefile,
+			"Corefile-backup": currentCoreDNSCorefile,
+		},
+	}); err != nil {
+		return errors.Wrap(err, "unable to update CoreDNS config map")
+	}
+
+	return nil
+}
+
+// UpdateCoreDNSDeployment will patch the deployment image to the
+// imageRepo:imageTag in the KCP dns. It will also ensure the volume of the
+// deployment uses the Corefile key of the coredns configmap.
+func (w *Workload) UpdateCoreDNSDeployment(ctx context.Context, dns *kubeadmv1.DNS) error {
+	image := fmt.Sprintf("%s:%s", dns.ImageRepository, dns.ImageTag)
+	patchJSON := fmt.Sprintf(
```

Comment:
This patch looks very similar to the one above, I'm wondering instead of dealing with string-interpolation if it'd make sense to refactor this a little in a different method and generate the path on demand depending on the input. This can come later though.